Add local cache for noderesourcetopology using the reserve plugin [1/3] - preparation

fromanirh opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Up until now, the `filter` and `score` plugin depend on the freshness of the noderesourcetopology (NRT) objects to make their decisions.
IOW, the plugins depend on timely reporting of NRT objects to learn about the actual resource availability; they consider the resources free until told otherwise.
This leads to resource overallocation, which in turn increases the likeness of TopologyAffinityError, which is what the plugins wants to avoid (or at least reduce as much as possible) in the first place.
    
 We call this behaviour `false positive`: the scheduler plugin wrongly  believes resources are available, while they are not, and schedule a pod while it should wait (re-enqueue for later).
 Similarly, we then define `false negative` the opposite behaviour, on which the scheduler plugin keeps the pod unscheduled (Pending) because it wrongly believes resources are NOT available.
    
This change introduces a scheduler side, NRT-aware cache leverage the `reserve` plugin, which tracks the pending resources, which the scheduler knows have been allocated to pods, but which aren't yet been reported as un-available by the NRT updater.
The cache is optional, simple as possible, requires minimal changes in the `filter` and `score` plugin, tolerates restarts well, and is backward compatible, meaning that `score` and `filter` should see NO CHANGES in their behaviour with the cache disabled.
On the other hand, with the cache enabled, the resources allocated to pods are decremented, hence the `filter` and `score` see a more realistic, albeit more pessimist (out of necessity) picture.
    
This has two main benefits:
1. we have now the option to greatly reduce the pressure on the topology updater, because good scheduling decision is less dependent on timely update of NRT objects (`reserve` of course reconciles its own state with fresh NRT state once available)
2. we add a knob to cluster administrators to prefer to keep pods in waiting state (false negatives using the the definition above) and have less frequent topology updates.
    
Overall, the NRT aware cache allows to optionally reduce the pressure on the topology update frequency while being fully backward compatible and optional.

#### Which issue(s) this PR fixes:
Fixes N/A

#### Special notes for your reviewer:
Depends on: #427 
The work is partitioned as follows:
- the initial PR #315 was expanded in scope as flaws in the original old design emerged during the review. We now use #315 to add documentation, context and the very basic groundwork.
- the followup #437 extends and completes #315. PRs are kept split to make the review easier (or to attempt so) and to preserve original context of #315 and all the discussion. However, we need both #315 and #437 (the latter bundles the former) to have the bare minimum required functionality
- finally, we have #439 which extends further #437. Changes added here aren't part of the minimal core but add features needed in real-world scenarios. This work can be reviewed separately and builds on top of #437. This completes the feature.

#### Does this PR introduce a user-facing change?
```release-note
Introduce optional cache for noderesourcetopology objects, depending both on the reserve plugin being enabled and on explicit configuration option.
```

fromanirh commented:
@AlexeyPerevalov  @swatisehgal  @Tal-or

fromanirh commented:
`reserve` is being tested, will lift the WIP as soon as I'm confident enough (= run some e2e tests, possibly add them)

cynepco3hahue commented:
I like the PR, it is much nicer to have pending pods than a lot of failure pods with the affinity error.

swatisehgal commented:
/ok-to-test

swatisehgal reviewed (commented):
Thanks for working on this @fromanirh! Design wise it looks good to me, just left a couple of minor comments. Appreciate you adding the unit tests, (as you have already planned) would be nice to have some e2e or integration tests for this well.

AlexeyPerevalov commented on pkg/noderesourcetopology/pluginhelpers.go:
```diff
@@ -17,49 +17,15 @@ limitations under the License.
-	return nodeTopology
-}
-
-func initNodeTopologyInformer(kubeConfig *restclient.Config) (listerv1alpha1.NodeResourceTopologyLister, error) {
```
I think you can keep initNodeTopologyInformer and call setupNodeResourceCountersWithInformer from there.

fromanirh replied:
Sure, I can change accordingly in the next upload.

fromanirh commented on pkg/noderesourcetopology/score.go:
```diff
@@ -55,7 +55,7 @@ func (rw resourceToWeightMap) weight(r v1.ResourceName) int64 {
 func (tm *TopologyMatch) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
 	klog.V(5).InfoS("Scoring node", "nodeName", nodeName)
-	nodeTopology := findNodeTopology(nodeName, tm.lister)
+	nodeTopology := tm.getNodeTopology(nodeName)
```
I think I handled this, let's check in the next upload.

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -17,9 +17,134 @@ limitations under the License.
+
+	informer := nodeTopologyInformer.Informer()
+	// After every update to NRTs, we are getting fresher information from the node topology updaters;
+	// this means we can safely discarc the pending data we stored, because the information is now consolidated
```
thanks, fixed!

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -17,9 +17,134 @@ limitations under the License.
+
+type nodeResourceCounters struct {
+	rwlock sync.RWMutex
+	nodes  map[string]resourceCounter
```
good idea, adding.

fromanirh commented:
> TODO: make sure we handle burst of deletions

No code change seems needed. Add a code comment to explain why.

fromanirh commented:
existing tests should now pass and cover the non-regression promise. I'll add tests to demonstrate and verify the new behaviour and then lift the WIP.

Huang-Wei commented on pkg/noderesourcetopology/resourcecounter.go:
```diff
@@ -0,0 +1,111 @@
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+)
+
+type resourceCounter v1.ResourceList
```
Is the purpose of aliasing v1.ResourceList to come up with a dedicated Add() and Sub() functions? If so, I think we can add a stateless function like:

- func Add(a v1.ResourceList, b v1.ResourceList) v1.ResourceList
- func Sub(a v1.ResourceList, b v1.ResourceList) v1.ResourceList

So that we can eliminating introducing this alias.

fromanirh replied:
sure, I'm fine with this and I actually like more the free function approach.

Huang-Wei commented on test/integration/noderesourcetopology_test.go:
```diff
@@ -600,6 +534,131 @@ func TestTopologyMatchPlugin(t *testing.T) {
+		t.Logf(" Node %s created: %v", nodeName, n)
+	}
+
+	nodeList, err := cs.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
```
The err is not used anywhere, which failed the integration test.

fromanirh replied:
thanks, bad rebase. Fixed! Going to add the actual test ASAP and then lift the WIP.

Huang-Wei reviewed (commented):
Thanks for introducing the Reserve plugin! Some comments below.

Huang-Wei commented:
@fromanirh I rethink this problem a bit: the NRT plugin's filter() logic relies on the NRT object's "Available" field:

https://github.com/kubernetes-sigs/scheduler-plugins/blob/7b3182b2e1750f673d00ae0edcb5a3609a528e4a/pkg/noderesourcetopology/pluginhelpers.go#L143

While this field gets frequently updated (by a NRT deamon component on node side I suppose), and thus the delay of the arriving NRT event may cause "false positive".

If my understanding is right, have you considered re-architecting this plugin to NOT rely on NRT's "Available" field; instead, maintain a NRTInfo (cache) in the plugin by add/update/delete/reserve/unreserve the pods, so that it'd only rely on the NRT object's "Allocable" info (which is generally static since initiation) and hence barely no "false positive".

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+type nodeResourceCounters struct {
+	rwlock sync.RWMutex
```
nice! changing

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -0,0 +1,187 @@
+
+}
+
+type nodeResourceCounters struct {
```
yes, this is probably better. Changing.

fromanirh commented on pkg/noderesourcetopology/plugin.go:
```diff
@@ -63,9 +63,11 @@ type TopologyMatch struct {
 	policyHandlers      PolicyHandlerMap
 	scorerFn            scoreStrategy
 	resourceToWeightMap resourceToWeightMap
+	pendingResources    *nodeResourceCounters
```
sure thing

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -0,0 +1,187 @@
+
+type nodeResourceCounters struct {
+	rwlock sync.RWMutex
+	// nodeName -> resourceCounter (alias for k8s.io/api/core/v1.ResourceList)
```
I agree that simplifying here is beneficial, and I'll implement your suggestions below.

fromanirh commented on pkg/noderesourcetopology/reserve.go:
```diff
@@ -0,0 +1,187 @@
+			klog.V(5).InfoS("Unexpected object")
+			return
+		}
+		nrc.Flush(nrt.Name)
```
The key assumption (and the intended behaviour) is that when a NRT update comes, being the update done by the agent running on the node which knows the exact allocation of resources, then we don't need the pessimistic approximation implemented by the `nrtCache`. In other words, if we have a streams of NRT updates NRT_1, NRT_2, ... NRT_N we have that NRT_(n) + nrtCache approximates (pessimistically) NRT_(n+1).
From the above, if we get an update in between this is actually beneficial and is expected to improve the scheduling decisions.

To put yet in another way, the nrtCache, and the reserve plugin which manages it, is expected to improve the filtering decision during the interval between two NRT updates. The interval can be variable (and not fixed) if the NRT updating agent on the nodes adopts a event-based implementation, but still we will have a interval.

Additionally, depending on the per-cluster tradeoff, having a nrtCache can allow the administrator to have longer NRT updating intervals, while previously longer NRT intervals almost always implied a decrease of the filtering/scheduling quality decisions due to less fresh data available to the scheduler.

Huang-Wei replied:
> The key assumption (and the intended behaviour) is that when a NRT update comes, being the update done by the agent running on the node which knows the exact allocation of resources

But can we really assume that? Let's say:

- at T1, pod1 gets scheduled and assumed on node X
- at T2, pod2 gets scheduled and assumed on node X
- at T3, the event of pod1's placement arrives at node, and then the NRT object gets updated. Note: only pod1's event arrives, not pod2.
- at T4, the NRT object's update arrives at the scheduler plugin, and the current logic removes the node entirely, i.e., invalidate pod2 from the assumed cache
- at T5, we try to schedule pod3, in this case, we're not aware of the resources reserved by pod2

swatisehgal replied:
This is a great point! We definitely need to handle a case like this properly here.

fromanirh replied:
@Huang-Wei you are very right. This is a part I overlooked. The cache invalidation needs to be smarter and more robust to handle cases like this. Thanks for raising this important point.

Huang-Wei commented:
> First, the authoritative source of data for Capacity/Allocatable/Available is the NRT updating agent on the nodes.

I'm not sure I like it. This adds extra responsibility to the agent, and to me, it's single point of failure - what if it doesn't work while kubelet works properly. The more key problem is: even if we go this approach, I think we cannot prevent "false positive", see https://github.com/kubernetes-sigs/scheduler-plugins/pull/315#discussion_r784310380.

I'd say "less is more", to offload the responsibility of the agent may pay more back. And this is how kubelet and scheduler interacts - kubelet doesn't update the node's Available resource. It's been working greatly for years.

swatisehgal commented:
> > First, the authoritative source of data for Capacity/Allocatable/Available is the NRT updating agent on the nodes.
> 
> I'm not sure I like it. This adds extra responsibility to the agent, and to me, it's single point of failure - what if it doesn't work while kubelet works properly. The more key problem is: even if we go this approach, I think we cannot prevent "false positive", see [#315 (comment)](https://github.com/kubernetes-sigs/scheduler-plugins/pull/315#discussion_r784310380).
> 
> I'd say "less is more", to offload the responsibility of the agent may pay more back. And this is how kubelet and scheduler interacts - kubelet doesn't update the node's Available resource. It's been working greatly for years.

I do like the idea of moving the accounting logic to the scheduler plugin and the fact that it aligns with how kubelet and scheduler interacts but at the same time I see a few problems with this approach.

Assuming we have a cluster with no workloads running initially, the way I see the flow is something like below:
1. NRT exposes static information corresponding to the node which includes Topology manager policy and the resource allocatable on a per NUMA basis
2. As suggested by @Huang-Wei above, we update the scheduler -plugin cache with   by add/update/delete/reserve/unreserve.
3. For a pod that needs to be scheduled, the scheduler-plugins updates the cache. We need to determine or make an assumption (based on the allocatable resource and TM policy) a possible NUMA node. I believe the cache itself needs to keep track of resources on a per NUMA basis. We might have to change `NRTInfo` cache to something like below maybe?:
```
type nrtCache struct {
	sync.RWMutex
	// nodeName -> NUMA node -> k8s.io/api/core/v1.ResourceList
	// hence we track all the pending resources for each node and numa node.
	nodes map[string]map[string]corev1.ResourceList
}
```
4. We need to keep in mind that the Topology Manager can decide to allocate resources from a completely different NUMA node but from scheduler-plugin's perspective determining the exact NUMA node is not important, what matters is whether a node is capable of fitting a pod on it or not.

Here are the issues I see with this approach:

1.  Because the scheduler plugin is deployed at runtime and if we only have static information in the NRT CR we don't have a clear picture of the previously running pods and thus the available resources at that point in time. A possible solution is to continue leveraging the evaluated (_available_ resources per NUMA) information in topology-updater/agent to have a clear starting point. We should use that information to populate the cache and then proceed further when a pod request comes.
2.  We need to make sure that `NRTInfo` cache a "close enough" representation of the reality. Consider a scenario where a pod after going through the evaluation process, updates the `NRTInfo` cache and is scheduled on a node but fails to start. (It might or might not be because of Topology Manager as there can be other reasons like failure to setup network or attach volumes)
In a similar vein, we need to take into consideration that allocatable resources can change during the lifecycle of kubelet, eg. deployment of a device plugin in the cluster. How do we propogate the information of a newly introduced device in a cluster to `NRTInfo` cache

So, it boils down to how we reconcile the actual state with the state captured in the scheduler plugin cache? IMO it is very important for us to address these if we were to go in this direction.

/cc @AlexeyPerevalov for his thoughts on this subject

Huang-Wei commented:
Thanks @swatisehgal for the detailed explanation.

Let me try to summarize: even with a detailed `nrtCache` (suggested in 3), it may also be inconsistent with the eventual resource allocation as "Topology Manager can decide to allocate resources from a completely different NUMA node" (mentioned in 4). If so, IMO the reserving idea is problematic: assumed pod (but not updated yet in NRT) may be subtracted **incorrectly** on a different NUMA slot?

swatisehgal commented:
Just want to clarify that the flow in my [comment](https://github.com/kubernetes-sigs/scheduler-plugins/pull/315#issuecomment-1013538180) above was corresponding to the approach you suggested where we offload the accounting responsibility from topology-updater/agent to scheduler plugin.
In point 4, what I was trying to convey was at a scheduler-plugin level we only need to maintain a "close-enough" state meaning we don't need information in terms of exactly which NUMA have the resources been allocated from i.e. 
We DON'T need something like this: Node A we have NUMA 0 that has 4 CPU, NUMA 1 that has 3 CPUs and so on. 
We ONLY need: Node A has two NUMA nodes and one has 3 CPUs and another one has 4 and based on _just_ that information we can make the decision whether that "node" is suitable or not.

If we need to offload the responsibility from the agent to the plugin we **HAVE** to nail down how we address the inconsistency in terms of the actual state (in cases like pod failure, device plugin introduction) and the real state.

swatisehgal commented:
I am conscious here that we are discussing a massive redesign by questioning some of the architectural decisions we had made right at the beginning so I apologize for the rant above and for now, I am going to try and focus on this PR itself :)  

When it comes to reserve plugin approach, I see the approach here valuable as we would prefer to have a pessimistic view of the world and have the pod wait a bit longer than make a wrong decision that results in failed pod. @fromanirh's approach to decrement resources from all NUMA nodes in the cache seems reasonable but concerns [here](https://github.com/kubernetes-sigs/scheduler-plugins/pull/315#discussion_r784310380) are valid! In order to address this issue, I think we need to have some intelligence in the cache to be able to take into consideration that when it comes to pods, the scheduler will to have the most up to date information (because it is the first point of contact) and when it comes to allocated resources, NUMA etc, kubelet will have the most up to date info. So if we take that into consideration when we reconcile the NRTInfo cache we should be able to address this problem? Let me explain how:

at T1, pod1 gets scheduled and assumed on node X
at T2, pod2 gets scheduled and assumed on node X
at T3, the event of pod1's placement arrives at node, and then the NRT object gets updated. Note: only pod1's event arrives, not pod2.
at T4, the NRT object's update arrives at the scheduler plugin, here rather than  invalidating and removing pod2 from the assumed cache we trust that the scheduler is ahead of the kubelet and we go ahead assuming pod2 exists. We allow for a certain interval of time for the updated info to be obtained from the NRT CR to reconcile the state from the NRT. If there is no update, we unreserve resources that we assumed should have been allocated to Pod2.

We will again have to be cautious in handling cases where the pod fails after getting scheduled on the node.

WDYT?

Huang-Wei commented:
> We ONLY need: Node A has two NUMA nodes and one has 3 CPUs and another one has 4 and based on _just_ that information we can make the decision whether that "node" is suitable or not.

I see your point: `adjustWithResourceCounters()` does the subtraction (from nrtCache) on **all** NUMA slots. This is sub-optimal, but can avoid false positives. In your case, suppose the assumed pod requests for 3 CPUs, the runtime allocation for NUMA1/NUMA2 could be one of the cases: {3/3, 0/4}, or {0/3, 3/4}. But we are not aware of that, so can only pessimistically subtract the 3 CPUs from **both** NUMA slots to be: {3/3, 3/4}. So until the scheduler plugin gets update from NRT client/controller that reflect the eventual NUMA resource allocation, a pod requesting 2 CPUs cannot be scheduled.

> I am conscious here that we are discussing a massive redesign by questioning some of the architectural decisions we had made right at the beginning so I apologize for the rant above (might be best to discuss the new design in a separate thread). I am going to try and focus on this PR for now :)

I'm not _suggesting_ a redesign, just explore the options. I like the discussion and I think it clarifies the problem a bit and help me understand the implications out there. I believe once it lifts up to k/k, you will also be asked similar questions :)

> When it comes to reserve plugin approach, I see the approach here valuable

Agree. We're on the same page now.

> but the concerns here are valid!

This is a thing we need to resolve. But I cannot think of an ideal solution as neither a NRT event tells us which pods are assigned, nor a pod update event guarantees the NRT obj is up-to-date by that time.

> at T4, the NRT object's update arrives at the scheduler plugin, here rather than invalidating and removing pod2 from the assumed cache we trust that the scheduler is ahead of the kubelet and we go ahead assuming pod2 exists.

We don't have enough info to tell if pod2 should be invalidated or not. A _close_ solution is to figure out the delta between NRT update, and see if it matches any assumed pod. But it's subtle as the delta may contain multiple pods, mixing their successful and failed assignments.

swatisehgal commented:
> > We ONLY need: Node A has two NUMA nodes and one has 3 CPUs and another one has 4 and based on _just_ that information we can make the decision whether that "node" is suitable or not.
> 
> I see your point: `adjustWithResourceCounters()` does the subtraction (from nrtCache) on **all** NUMA slots. This is sub-optimal, but can avoid false positives. In your case, suppose the assumed pod requests for 3 CPUs, the runtime allocation for NUMA1/NUMA2 could be one of the cases: {3/3, 0/4}, or {0/3, 3/4}. But we are not aware of that, so can only pessimistically subtract the 3 CPUs from **both** NUMA slots to be: {3/3, 3/4}. So until the scheduler plugin gets update from NRT client/controller that reflect the eventual NUMA resource allocation, a pod requesting 2 CPUs cannot be scheduled.

What I am suggesting is slightly different from @fromanirh's current approach. In the example you mentioned, where a pod requests for 3 CPUs, rather than pessimistically subtracting resources from all numa nodes we make an educated guess using TM policy/CPU Manager policy (we know that Topology Manager with single-numa-node policy and CPU manager with static policy and no policyOptions try to allocate resources sequentially and then use that information) to determine the NUMA node where it is highly likely that resources will be allocated from. Say NUMA1 was our educated guess, then rather than subtracting resource request from both NUMA nodes we decrement from only one NUMA node in the `nrtCache`  which would look like  {0/3, 4/4} and in that case a pod with 2 CPUs can be scheduled. We later double check if our assumption was correct when the reconciliation happens and NRT CR is updated select. The number of NUMA nodes impact the probability of us selecting the right NUMA node but remember that we are making an educated guess (which if wrong, can be corrected later; I have explained this in my last point).
> I'm not _suggesting_ a redesign, just explore the options. I like the discussion and I think it clarifies the problem a bit and help me understand the implications out there.

Great, I like the discussion too! Happy that you found it helpful, I was worried I was digressing. 
>  I believe once it lifts up to k/k, you will also be asked similar questions :)

This is very true!

> > but the concerns here are valid!
> This is a thing we need to resolve. But I cannot think of an ideal solution as neither a NRT event tells us which pods are assigned, nor a pod update event guarantees the NRT obj is up-to-date by that time.
I understand, this is a tough one! I need to think about it a bit more as well
> > at T4, the NRT object's update arrives at the scheduler plugin, here rather than invalidating and removing pod2 from the assumed cache we trust that the scheduler is ahead of the kubelet and we go ahead assuming pod2 exists.
> 
> We don't have enough info to tell if pod2 should be invalidated or not. A _close_ solution is to figure out the delta between NRT update, and see if it matches any assumed pod. But it's subtle as the delta may contain multiple pods, mixing their successful and failed assignments.

Yeah, I came to this realization as well over the weekend :) I was pretty much thinking on the same lines: exposing some sort of diff or events (which will be more straightforward with podresource watch support) in the NRT CRs that gives more information to the scheduler plugin. I think if we have more granular information we can avoid the problem of "a delta containing multiple pods, mixing their successful and failed assignments". 

Kind of like how events are exposed in case of deployments, the we capture events corresponding to add/update/delete events of pods. It might look something like this:
```
  Events:
    Type    Reason         Age       From               Pod Name        Message
    ----    ------       ------     -------              -------        -------
    Normal  AddEvent     2m    nfd-topology-updater    pod1        Pod: podA, namespace: TestNs with resource request  {cpu: 2, memory: 64Mi } Added
    Normal  AddEvent     4m    nfd-topology-updater    pod3        Pod: podB, namespace: TestNs with resource request  {cpu :2, memory: 64Mi } Added
```
So, in case the scheduler-plugin scheduled pod1, pod2 and pod3 (and updated `nrtCache`)  we know that Pod2 didn't run on the node so we need to know now that pod2 probably failed and update the cache accordingly.

fromanirh commented:
Apologies for joining the discussion a bit late, I had a couple hectic days.
First of all thanks @swatisehgal for the brilliant summarization; you described the problem pretty much like I would have done, kudos. Your description is indeed fine.

> > We ONLY need: Node A has two NUMA nodes and one has 3 CPUs and another one has 4 and based on _just_ that information we can make the decision whether that "node" is suitable or not.
> 
> I see your point: `adjustWithResourceCounters()` does the subtraction (from nrtCache) on **all** NUMA slots. This is sub-optimal, but can avoid false positives. In your case, suppose the assumed pod requests for 3 CPUs, the runtime allocation for NUMA1/NUMA2 could be one of the cases: {3/3, 0/4}, or {0/3, 3/4}. But we are not aware of that, so can only pessimistically subtract the 3 CPUs from **both** NUMA slots to be: {3/3, 3/4}. So until the scheduler plugin gets update from NRT client/controller that reflect the eventual NUMA resource allocation, a pod requesting 2 CPUs cannot be scheduled.

Yes. In general, no matter how good the logic is on the scheduler side[1], the accounting of the resources can only be speculative, meaning that the topology manager on the kubelet is free to pick a different NUMA node that the one the scheduler picked, as long as the allocation requirements are fullfilled. And we can't predict this, we can only reconcile. Hence the reason why I started with a pessimistic allocation.

So, in the current architecture we will still need _a_ form of accounting on the node side, reflecting the actual allocation decision made by the Topology Manager.

[1] we can go as far as possibly reusing the same packages from the kubelet. Not I'm **NOT** proposing this anytime soon, we would need a lot of cleanups and refactoring to ge there; I mean however that this is the best we can theoritically do.

> > I am conscious here that we are discussing a massive redesign by questioning some of the architectural decisions we had made right at the beginning so I apologize for the rant above (might be best to discuss the new design in a separate thread). I am going to try and focus on this PR for now :)
> 
> I'm not _suggesting_ a redesign, just explore the options. I like the discussion and I think it clarifies the problem a bit and help me understand the implications out there. I believe once it lifts up to k/k, you will also be asked similar questions :)

I'm glad we are having this conversation which I find very helpful and I'm very glad to experiment further in this area!
The only suggestion I have is to keep the threads clearly separated, which we are already doing anyway :)

> > but the concerns here are valid!
> 
> This is a thing we need to resolve. But I cannot think of an ideal solution as neither a NRT event tells us which pods are assigned, nor a pod update event guarantees the NRT obj is up-to-date by that time.

Absolutely. I will need to re-evaluate the cache invalidation design, and of course discuss it. This is my top priority atm for this work, alongside the missing integration tests (already in progress though).

> > at T4, the NRT object's update arrives at the scheduler plugin, here rather than invalidating and removing pod2 from the assumed cache we trust that the scheduler is ahead of the kubelet and we go ahead assuming pod2 exists.
> 
> We don't have enough info to tell if pod2 should be invalidated or not. A _close_ solution is to figure out the delta between NRT update, and see if it matches any assumed pod. But it's subtle as the delta may contain multiple pods, mixing their successful and failed assignments.

Yes, I think there is no low-hanging fruit here. This would need some more thinking from my side.

fromanirh commented:
> > > We ONLY need: Node A has two NUMA nodes and one has 3 CPUs and another one has 4 and based on _just_ that information we can make the decision whether that "node" is suitable or not.
> > 
> > 
> > I see your point: `adjustWithResourceCounters()` does the subtraction (from nrtCache) on **all** NUMA slots. This is sub-optimal, but can avoid false positives. In your case, suppose the assumed pod requests for 3 CPUs, the runtime allocation for NUMA1/NUMA2 could be one of the cases: {3/3, 0/4}, or {0/3, 3/4}. But we are not aware of that, so can only pessimistically subtract the 3 CPUs from **both** NUMA slots to be: {3/3, 3/4}. So until the scheduler plugin gets update from NRT client/controller that reflect the eventual NUMA resource allocation, a pod requesting 2 CPUs cannot be scheduled.
> 
> What I am suggesting is slightly different from @fromanirh's current approach of updating `adjustWithResourceCounters()`. In the example you mentioned, where a pod requests for 3 CPUs, rather than pessimistically subtracting resources from all numa nodes we make an educated guess using TM policy/CPU Manager policy (we know that Topology Manager with single-numa-node policy and CPU manager with static policy and no policyOptions try to allocate resources sequentially and then use that information) to determine the NUMA node where it is highly likely that resources will be allocated from. Say NUMA1 was our educated guess, then rather than subtracting resource request from both NUMA nodes we decrement from only one NUMA node in the `nrtCache` which would look like {0/3, 4/4} and in that case a pod with 2 CPUs can be scheduled. We later double check if our assumption was correct when the reconciliation happens and NRT CR is updated select. The number of NUMA nodes impact the probability of us selecting the right NUMA node but remember that we are making an educated guess (which if wrong, can be corrected later; I have explained this in my last point).

I'm very interested in exploring this option. I think that we will be able to build on the foundation laid down in this PR.
My biggest concern is about reimplementing more of the topology manager logic in the scheduler plugin.
I think we would need to be careful about where to draw the line, and to evaluate the tradeoffs carefully.
Which is something I'd be happy to! :) I guess a deciding factor would be to be able to make this extension opt-in and configurable, so it's easier to evaluate the scheduling decisions and compare.

swatisehgal commented:
> Apologies for joining the discussion a bit late, I had a couple hectic days. First of all thanks @swatisehgal for the brilliant summarization; you described the problem pretty much like I would have done, kudos.

Thanks @fromanirh, happy to help! This has been a super interesting conversation so won't complain :) 

I agree with your assessments of there being no low hanging fruit unfortunately but lets keep churning our brains and we can have another round of discussion.

> 
> I'm very interested in exploring this option. I think that we will be able to build on the foundation laid down in this PR.
> My biggest concern is about reimplementing more of the topology manager logic in the scheduler plugin.
> I think we would need to be careful about where to draw the line, and to evaluate the tradeoffs carefully.
> Which is something I'd be happy to! :) I guess a deciding factor would be to be able to make this extension opt-in and configurable, so it's easier to evaluate the scheduling decisions and compare.

I agree we have to strike a balance and not make the scheduler too "intelligent" (can't believe I am saying that!).
A big +1 on making this opt-in and am looking forward to seeing this in action!

Huang-Wei commented:
Good conversations and now we have two problems upfront:

1. strategy to reserve pod resources - either pessimistically sub from all NUMA nodes, or intelligently sub from the most-likely NUMA node
I'm slightly inclined to the former one as IMO "false negative" is way better than "false positive". The upstream is also adopting this like ResourceQuota admission. However, if we can find a way to determine the NUMA node in the latter case, I'm fine with that.
2. invalidate pods properly
My 2cents before we find a perfect solution: make the logic a bit pessimistic. That's to say, only invalidate the pod when we're 100% sure; otherwise don't invalidate the pod.

fromanirh commented:
> Good conversations and now we have two problems upfront:
> 
>     1. strategy to reserve pod resources - either pessimistically sub from all NUMA nodes, or intelligently sub from the most-likely NUMA node
>        I'm slightly inclined to the former one as IMO "false negative" is way better than "false positive". The upstream is also adopting this like ResourceQuota admission. However, if we can find a way to determine the NUMA node in the latter case, I'm fine with that.
> 
>     2. invalidate pods properly
>        My 2cents before we find a perfect solution: make the logic a bit pessimistic. That's to say, only invalidate the pod when we're 100% sure; otherwise don't invalidate the pod.

Thanks for the summary. I agree with both. I'll be focusing on point 2 in the coming days. Since I'm kinda back to the drawing board it will take me a bit to update the PR, I'll keep you folks posted. Thanks for the great inputs!

fromanirh commented:
Over the past months I (with help from my team) redesigned the logic. The PR now contains the new logic and is reviewable. 
WIP is kept because I want to add integration tests in this PR and run more tests on my side.
There is a more detailed [design doc](https://github.com/fromanirh/fromanirh/blob/main/docs/noderesourcetopology-cache/NUMA-aware-scheduler-side-reserve-plugin.md) (and [general documentation](https://github.com/fromanirh/fromanirh/tree/main/docs/noderesourcetopology-cache)) available.

fromanirh commented:
/remove-lifecycle stale

fromanirh commented:
Note for reviewers: I recently (couple days ago) resumed the work full throttle here. I'm rewriting the impl from scratch addressing all the comments, and there are related fixes along the way. Not worth reviewing while marked WIP.

fromanirh commented:
after a long while, finally updated.
I'm partially retargeting this PR, as the scope had to be increased to address what emerged during the first review.
In this PR, I'm laying the groundwork for the real addition of the scheduler-side cache, refactoring where needed, adding config parameters and the interfaces.

I'm also adding the design doc which explain why I believe we need a somehow novel approach and all the design decisions.

These changes are pretty stable. The cache implementation proper is ongoing, under heavy testing and tuning. I'll post the code in the coming weeks, timeline depends on the available capacity and how the tests are going.

Depends on: https://github.com/kubernetes-sigs/scheduler-plugins/pull/427

swatisehgal commented:
Thanks for your work on this, it is looking good so far to me.

swatisehgal commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+
+![mismatch](images/reserve6-mismatch.png)
+
+![update](images/reserve7-update.png)
```
We should update the Fingerprint for Node 2 to `Fingerprint: Y2` and update the scheduled pod details (same mistake as above).

fromanirh replied:
updated as well

fromanirh commented:
> Took a deeper look at the diagrams and noticed some of my mistakes. I am here to blame and apologize for the inconvenience.
That's no problem at all! Nothing to apologize for and surely no blame to take. I thank you for catching this mistake (which I don't think it's very severe and surely not at this stage), for your prompt correction and for your design contributions and review. If there's anythinbg to improve, is that I promise more careful reviews myself next time. Thanks again for your work and contributions! I'll review and apply your corrections on my next upload.

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+resource availability listing the NRT objects on the cluster and
+populates its own cache using this data.
+
+![watch](images/reserve3-watch.png)
```
thnaks, updated

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+from the cache. NRT objects are no longer considered to compute the free
+resources (see below).
+
+![allocation](images/reserve5-allocation.png)
```
thanks, updated

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+    b.  If they match, overwrite the node cache with the content of the
+        > NRT object. The node cache is now clean again
+
+![mismatch](images/reserve6-mismatch.png)
```
thanks, updated

Tal-or commented on pkg/noderesourcetopology/cache/cache.go:
```diff
@@ -0,0 +1,75 @@
+	listerv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/generated/listers/topology/v1alpha1"
+)
+
+type Cache interface {
```
What is the advantage of an interface here over a struct which has those receiver functions?

fromanirh replied:
in this commit, no advantage. There's no need for interface and empty methods with this commit. In the following commit I'm adding another cache implementation, and the interface allows for transparent replacement and to allow to disable the cache functionality completely.

Tal-or reviewed (commented):
A couple of small comments from my side but overall LGTM

I'll review the other parts in the following days

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+    b.  If they match, overwrite the node cache with the content of the
+        > NRT object. The node cache is now clean again
+
+![mismatch](images/reserve6-mismatch.png)
```
Uhm, the path looks correct and the "show file" functionality displays the image correctly, so I can't say. I can't rule out a UI glitch.

Huang-Wei replied:
nit: is there a typo in this picture that it should be "Y1" in the left bottom picture?

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+
+![mismatch](images/reserve6-mismatch.png)
+
+![update](images/reserve7-update.png)
```
same

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+
+At node level, a deamonset enables components which query the resource
+allocation state from the kubelet - which is the ultimate source of
+truth -, reorganize this data with per-NUMA node granularity, fills and
```
rephrased completely, thanks.

fromanirh commented on pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md:
```diff
@@ -0,0 +1,340 @@
+changes), or with a mixed approach.
+
+Before each scheduling decision, the NUMA-aware scheduler plugins list
+the possible NRT objects and perform their decisions on the available
```
thanks, rephrased.

swatisehgal commented:
Thanks for addressing all my review comments and updating the diagrams. I am happy enough with the proposed design, we had pretty much agreed on adopting a pessimistic approach previously and I see that captured as part of the design doc.

It would be great to hear @Huang-Wei's thoughts on the cache invalidation approach proposed here. We need to compare the NRT state to that of the cache to determine if the NRT is upto date and reflects the correct state of the system so that the reserve plugin cache can be flushed. Podfingerprint seems like a nice approach to me to compare NRT states as just comparing the cache and NRT objects is not enough, we need to compare the states when both the cache and NRT objects are describing the same state of the system wrt the pods running on those nodes and we need to uniquely represent and determine that in some way. 

/lgtm 

/assign @Huang-Wei 
for review/approval

Huang-Wei reviewed (commented):
Given I haven't reviewed follow-ups, some comments may not be true.

Huang-Wei commented:
> It would be great to hear @Huang-Wei's thoughts on the cache invalidation approach proposed here.

Thanks for the neat doc and pictures. Looks pretty promising.

My question is: is the NRT updater agent going to update the NRT status periodically? If so, my concern is it may not work well with pod updates to come up with an accurate picture of NRT usage globally:

- we have a Node with resource ✅`{numa1: 4/4cpu, numa2: 4/4cpus}`
- pod1 and pod2 requesting for 1cpu and 2cpus, and gets assigned to numa1 and numa2 accordingly
- so now the NRT status is ✅`{numa1: 1/4cpu, numa2: 2/4cpus}`
- pod3, pod4 come in and both requests for 2cpus and 1cpu, and they get assumed internally as ⚠️`{numa1: 4/4cpu, numa2: 4/4cpus}`
- T1: NRT updater agent catches up and issues an update with NRT status ✅`{numa1: 4/4cpu, numa2: 2/4cpus}`
- So how can we invalidate the cache properly on the following 3 scenarios?
    1. pod3 update received before T1, pod4 update received after T1
    2. pod3 and pod4 updates received before T1
    3. pod3 and pod4 updated received after T1

Huang-Wei commented:
I'm not sure it's possible to have the NRT update agent to just issue event on "which pod assigned/deleted on which numa slot", in that case, the reserve and invalidation logic will be pretty straightforward:

- reservation logic works unchanged
- invalidation logic will only watch the pod-assigned/deleted-on-which-numa slot event, and unreserve the resources properly
- (the internal cache needs to have the data structure recoding which slots are over-reserved, so it can be deducted later properly)

fromanirh commented on pkg/noderesourcetopology/cache/cache.go:
```diff
@@ -0,0 +1,75 @@
+	// this sequence of events as the previous pod required too much - a possible and benign condition.
+	ReserveNodeResources(nodeName string, pod *corev1.Pod)
+
+	// ReleaseNodeResources decrement from the node assumed resources the resources required by the given pod.
+	ReleaseNodeResources(nodeName string, pod *corev1.Pod)
```
I used `Release` because it has the same character count of `Reserve' :) but `Unreserve` is clearer, will change accordingly.

fromanirh commented on pkg/noderesourcetopology/cache/cache.go:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+func (pt Passthrough) GetByNode(nodeName string, _ *corev1.Pod) *topologyv1alpha1.NodeResourceTopology {
```
this will be used in upcoming commits, by the `overallocating` cache implementation (https://github.com/kubernetes-sigs/scheduler-plugins/pull/437/commits/c5f65867bdd22933213024e334a4ca35b4d8f804). I acknowledge it is awkward to add stuff we don't use, but I'm struggling to split the huge PR in manageable chunks.

fromanirh commented on apis/config/v1beta3/types.go:
```diff
@@ -143,7 +143,10 @@ type ScoringStrategy struct {
+	// ScoringStrategy a scoring model that determine how the plugin will score the nodes.
 	ScoringStrategy *ScoringStrategy `json:"scoringStrategy,omitempty"`
+	// If > 0, enables the caching facilities of the reserve plugin - which must be enabled
+	CacheResyncPeriodSeconds *int64 `json:"cacheResyncPeriodSeconds,omitempty"`
```
It will be used in followup to implement the reconciliation logic: https://github.com/kubernetes-sigs/scheduler-plugins/pull/437/commits/c5f65867bdd22933213024e334a4ca35b4d8f804#diff-4ad175c1f91ec022ebcea71714a329e01f594c0ac6487650b3d779eb33e08f41R164

fromanirh commented on pkg/noderesourcetopology/pluginhelpers.go:
```diff
@@ -20,50 +20,38 @@ import (
 
-func initNodeTopologyInformer(kubeConfig *restclient.Config) (listerv1alpha1.NodeResourceTopologyLister, error) {
-	topoClient, err := topoclientset.NewForConfig(kubeConfig)
+func initNodeTopologyInformer(tcfg *apiconfig.NodeResourceTopologyMatchArgs, handle framework.Handle) (nrtcache.Cache, error) {
```
yep, using in followups. If you want, I'm all for more convenient split of the contribution, but no real nice split emerged so far :\

fromanirh commented:
> My question is: is the NRT updater agent going to update the NRT status periodically?

The updater can update periodically or when pods get created/deleted (using CRI hooks). But for scalability reasons we want to throttle the updates sent to the apiserver, which
severely limits the usefulness of event-based updates. It's a complex set of tradeoffs.

> If so, my concern is it may not work well with pod updates to come up with an accurate picture of NRT usage globally:
> we have a Node with resource white_check_mark{numa1: 4/4cpu, numa2: 4/4cpus}
> pod1 and pod2 requesting for 1cpu and 2cpus, and gets assigned to numa1 and numa2 accordingly
> so now the NRT status is white_check_mark{numa1: 1/4cpu, numa2: 2/4cpus}
> pod3, pod4 come in and both requests for 2cpus and 1cpu, and they get assumed internally as warning{numa1: 4/4cpu, numa2: 4/4cpus}

The catch is how we do the resync (aka "reconciliation", using the two terms interchangeably).
After both pod1 and pod2 are assigned and running:
- BEFORE the resync NRT will be {numa1: 1/4cpus, numa2: 1/4cpus} (notation lists *free* cpus over the total) because requests of cpus (1 and 2 respectively) will be accounted pessimistically to both NUMA nodes
- AFTER the resync NRT will be {numa1: 3/4cpus, numa2: 2/4cpus}

The resync is asynchronous wrt the scheduling queue and driven by the scheduler itself*

> T1: NRT updater agent catches up and issues an update with NRT status white_check_mark{numa1: 4/4cpu, numa2: 2/4cpus}

IIUC this means the updater will report the correct allocation, hence {numa1: 3/4cpus, numa2: 2/4cpus}, right?

> So how can we invalidate the cache properly on the following 3 scenarios?

In general: the scheduler (the overallocating aka pessimistic cache in the scheduler) will watch all the NRT updates, but will consider the updates
only after a succesfull resync. It will keep operating with its own pessimistic view up until a succesfully resync happens.
So let's call T1 the moment a succesfull resync happens from now on.

> pod3 update received before T1, pod4 update received after T1

pod3 will be processed using the pessimistic overallocation, which is {numa1: 1/4cpus, numa2: 1/4cpus}.
pod3 requires 2cpus, so scheduling is impossible: the node will be filtered out. We want this scheduling to be retried later.
pod4 will be processed using the real allocation state which is {numa1: 3/4cpus, numa2: 2/4cpus}.
pod4 requires 1 cpu, so it will be scheduled.
The new pessimistic overallocation will be {numa1: 2/4cpus, numa2: 1/4cpus}.
pod3 fits already using the pessimistic overallocation, so it will be scheduled.
The new pessmistic view could look like {numa1: 2/4cpus, numa2: 0/4cpus}. The resync loop will catch later.

In this case, pod4 can get scheduled before pod3. We don't consider this a problem, but I'm highlighting this just in case.

> pod3 and pod4 updates received before T1

pod3 will be processed using the pessimistic overallocation, which is {numa1: 1/4cpus, numa2: 1/4cpus}.
pod3 requires 2cpus, so scheduling is impossible: the node will be filtered out.
pod4 will be processed using, but it will fit! it will be scheduled, and the new pessimistic overallocation will be {numa1: 0/4cpus, numa2: 0/4cpus}.
pod3 will be now racing (and be pending - our goal is to keep it pending here!) until the resync catches up, and will be scheduled afterwards.

Also in this case, pod4 can get scheduled before pod3. We don't consider this a problem, but I'm highlighting this just in case.

> pod3 and pod4 updated received after T1

pod3 will be processed using the real allocation state which is {numa1: 3/4cpus, numa2: 2/4cpus}.
pod3 requires 2cpus, so it fits, it will be scheduled and the new pessimistic overallocation will look like {numa1: 1/4cpus, numa2: 0/4cpus}
pod4 will be processed using the just computed pessimistic overallocation, which is {numa1: 1/4cpus, numa2: 0/4cpus}.
pod4 requires 1cpus, so it fits! It will be scheduled and the new pessimistic overallocation will look like {numa1: 0/4cpus, numa2: 0/4cpus}

The resync loop will eventually catch up and pick the correct allocation state.

fromanirh commented:
> I'm not sure it's possible to have the NRT update agent to just issue event on "which pod assigned/deleted on which numa slot", in that case, the reserve and invalidation logic will be pretty straightforward:

We are exploring event-based notification on node side leveraging [OCI runtime hooks](https://github.com/opencontainers/runtime-spec/blob/main/config.md#posix-platform-hooks). There are two main problems however:

- deploying the hooks themselves is a challenge 
- for scalability reasons, we want to move carefully to avoid to generate events - thus a NRT object update after every pod lifecycle events. We don't want to overstress the apiserver.

>     * reservation logic works unchanged
> 
>     * invalidation logic will only watch the pod-assigned/deleted-on-which-numa slot event, and unreserve the resources properly
> 
>     * (the internal cache needs to have the data structure recoding which slots are over-reserved, so it can be deducted later properly)

This is true and nice approach. After some initial exploration, we were not too happy to expose the mapping between NUMA zones and pods. We are not sure about exposing this level of detail, and how to report this information: UPDATE-ing each pod for an annotation, for example, would still cause high traffic to the apiserver, which we are reluctant to do.

That said, this is still a direction we want to explore for future refinements.

fromanirh commented:
I'd like to add we are building a [operator to enable the noderesourcetopology plugin on openshift](https://github.com/openshift-kni/numaresources-operator) which comes with a [extensive and still growing e2e suite](https://github.com/openshift-kni/numaresources-operator/tree/main/test/e2e/serial) and we're seeing very promising results with this caching mechanism enabled.
We will be very happy to contribute back (perhaps a subset of) this suite back to this project (xref: https://github.com/kubernetes-sigs/scheduler-plugins/pull/349) in the future. The "only" challenges are to find good ways to deploy the stack and to make sure we have beefy enough machines in CI (obviously the e2e suite of this plugin needs to choose among at least 2+ nodes each with 2+ NUMA nodes)

Huang-Wei commented:
I understand how the over-reservation works inside NRT cache. My biggest concern is that the resync may not handle all cases to _always_ invalidate the cache correctly.

Let me ask a question about the following consequences:

1. P1 (requests for 1 CPU) is scheduled
2. P2 (requests for 2 CPUs) is scheduled
3. NRT update event, but with only P1's info
4. P1 gets deleted

IIUC, the result of NRT cache is like: (X/Y means X out Y is available)

- by the end of step 1, it's `{numa1: 3/4, numa2: 3/4}`
- by the end of step 2, it's `{numa1: 1/4, numa2: 1/4}`
- by the end of step 3, (suppose P1 is allocated to numa1), is it synced to `{numa1: 3/4, numa2: 4/4}`?
- by the end of step 4, is it deducting P1 to be `{numa1: 4/4, numa2: 4/4}`?
- if the above two hypotheses are correct, does it mean we may incorrectly schedule a Pod that requests 3 or 4 CPUs?

fromanirh commented:
> I understand how the over-reservation works inside NRT cache. My biggest concern is that the resync may not handle all cases to always invalidate the cache correctly.
> 
> Let me ask a question about the following consequences:
>
>    P1 (requests for 1 CPU) is scheduled
>    P2 (requests for 2 CPUs) is scheduled
>    NRT update event, but with only P1's info
>    P1 gets deleted

Thanks for raising the point about handling deletions.
With the caveat with we are still testing the deletion scenarios (while we have already more testing done in the addition scenarios),
let me explain how we plan to handle them.

The first noteworthy thing is that in the scenario above the scheduler cache will *receive* (think git stash) but not *consume* (think git commit) the NRT update.
Why?

Because the NRT plugin processed and scheduled _both_ P1 and P2, so it knows the node must now hold both P1 and P2, unless of course an event like a pod deletion or a rejection from the kubelet happened. This is not the case. So partial NRT updates will be received (the plugin is WATCH'ing NRT objects) but the internal cache will *not* be updated until the expected state is received. (note we are likely to need PRs in the very near future to make sure we detect node rejects like TopologyAffinityError, but this is the design and the intended behavior nevertheless)

Can we actually get partial updates? sure! the key is to let the node settle and the scheduler catch up. IOW, the resync mechanism is trying to ignore partial updates during bursts. In this scenario:

    P1 (requests for 1 CPU) is scheduled
    (some time passes)
    NRT update event, but with P1's info
    scheduler expects NRT update including P1 (for simplicity sake: with ONLY P1), so it commits this update to its own cache

In the above (for clarity's sake, redundant example)

    P1 (requests for 1 CPU) is scheduled
    P2 (requests for 2 CPUs) is scheduled
    NRT update event, but with only P1's info - but the scheduler expects a update which INCLUDES P1 and P2, so this is ignored.
      the scheduler keeps with its own pessimistically computed state

To compute the expected state, we use this approach: https://github.com/k8stopologyawareschedwg/podfingerprint
(note we do this now, exposing the mapping of pods<->numaZone on each compute node like discussed a couple commets ago is still on the table for future enhancements)

> IIUC, the result of NRT cache is like: (X/Y means X out Y is available)
>
>    by the end of step 1, it's {numa1: 3/4, numa2: 3/4}
>    by the end of step 2, it's {numa1: 1/4, numa2: 1/4}
>    by the end of step 3, (suppose P1 is allocated to numa1), is it synced to {numa1: 3/4, numa2: 4/4}?
>    by the end of step 4, is it deducting P1 to be {numa1: 4/4, numa2: 4/4}?
>    if the above two hypotheses are correct, does it mean we may incorrectly schedule a Pod that requests 3 or 4 CPUs?

No, because we ignore the partial updates during bursts as explained above (but I'll be happy to elaborate further).
Additionally, we intentionally DO NOT DO resource counters update on deletion. IOW, when the scheduler receives a pod deletion
event (we need to WATCH pods anyway) it just updates the set of expected pods in the node, which is then used by resync loop to
learn when the internal cache must be invalidated.
We don't change resources on the spot on delete, but we rather defer to the resync loop, exactly for the reason you highlighted:
the scheduler does NOT know the pod<->numaZone mapping, so we don't know which NUMA zone should get more free resources.

Huang-Wei commented:
@fromanirh thanks for the detailed explanation! Although a bit complicated, I think this can theoretically guarantee the correctness of cache invalidation.

To roll it forward, do you want to merge this PR first? (I may only have time to review the other 2 PRs starting tomorrow - today is code freeze for k8s 1.26)

fromanirh commented:
> @fromanirh thanks for the detailed explanation! Although a bit complicated, I think this can theoretically guarantee the correctness of cache invalidation.
> 
> To roll it forward, do you want to merge this PR first? (I may only have time to review the other 2 PRs starting tomorrow - today is code freeze for k8s 1.26)

Thank you @Huang-Wei fer the great reviews and your insights which led to a stronger (albeit unfortunately more complex) design.

It would be great to merge this PR first to make managing the other easier (rebases and so forth).

This one is intentionally super safe to merge on its own, because it only adds preparation steps.
We can review the other two later. (I have pending commits on part 2 of 3 anyway)

Huang-Wei commented:
@fromanirh thanks!

/lgtm
/approve
